src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -2,175 +2,181 @@
-        state = new HashMap<>(initialState);
-        this.index = index;
+    public int index() {
+        return withReadLockSupply(() -> index);
     }
 
-    public Snapshot clone() {
-        return new Snapshot(state, index);
+    /**
+     * @return a modifiable copy of this.
+     */
+    public Snapshot copySnapshot() {
+        return withReadLockSupply(() -> new Snapshot(state, index, false));
     }
 
     public Long getBalance(Hash hash) {
-        Long l;
-        rwlock.readLock().lock();
-        l = state.get(hash);
-        rwlock.readLock().unlock();
-        return l;
+        return withReadLockSupply(() -> state.get(hash));
     }
 
-    public Map<Hash, Long> patchedDiff(Map<Hash, Long> diff) {
-        Map<Hash, Long> patch;
-        rwlock.readLock().lock();
-        patch = diff.entrySet().stream().map(hashLongEntry ->
-            new HashMap.SimpleEntry<>(hashLongEntry.getKey(), state.getOrDefault(hashLongEntry.getKey(), 0L) + hashLongEntry.getValue())
-        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        rwlock.readLock().unlock();
-        return patch;
+    Map<Hash, Long> patchedDiff(Map<Hash, Long> diff) {
+        return withReadLockSupply(() -> {
+            Map<Hash, Long> map = new HashMap<>(diff.size());
+            diff.forEach((key, value) -> {
+                Long previous = state.getOrDefault(key, 0L);
+                map.put(key, Math.addExact(previous, value));
+            });
+            return map;
+        });
     }
 
     void apply(Map<Hash, Long> patch, int newIndex) {
-        if (!patch.entrySet().stream().map(Map.Entry::getValue).reduce(Math::addExact).orElse(0L).equals(0L)) {
-            throw new RuntimeException("Diff is not consistent.");
+        if (immutable) {
+            throw new UnsupportedOperationException("This snapshot is marked immutable. Make a copy if you want the 'apply' function;");
         }
         rwlock.writeLock().lock();
-        patch.entrySet().stream().forEach(hashLongEntry -> {
-            if (state.computeIfPresent(hashLongEntry.getKey(), (hash, aLong) -> hashLongEntry.getValue() + aLong) == null) {
-                state.putIfAbsent(hashLongEntry.getKey(), hashLongEntry.getValue());
+        try {
+            Long sum = patch.entrySet().stream().map(Map.Entry::getValue).reduce(Math::addExact).orElse(0L);
+            if (sum != 0L) {
+                throw new RuntimeException("Diff is not consistent. Sum is: " + sum);
             }
-        });
-        index = newIndex;
-        rwlock.writeLock().unlock();
-    }
 
-    public static boolean isConsistent(Map<Hash, Long> state) {
-        final Iterator<Map.Entry<Hash, Long>> stateIterator = state.entrySet().iterator();
-        while (stateIterator.hasNext()) {
+            patch.forEach((key, val) ->
+                state.compute(key, (k, previous) -> Math.addExact(val, previous != null ? previous : 0L)));
```

Comment:
For this one, I changed to `compute` because it is faster since internally `compute` keeps a hold of the Map.Entry and just changes the value.